Fix KiNET Port Out

DaAwesomeP opened the pull request:
OK, after looking at the outputs of QuickPlay Pro 1 and QuickPlay Pro 2, I found some things:

- I added the packet counter.
  - It is a big-endian counter that increments for each packet sent. It is not a frame counter, as it increments for each port packet that is sent to a device.
  - I don't know if it is a global counter across all devices for one KiNET source/generator or per-device, but I can test that next week. It probably doesn't matter since the devices wouldn't be able to tell anyway (unless it is used to check for missed packets and then it would need to be per-device).
  - I saw this counter climb to at least 3 bytes, but it is probably 4 bytes since otherwise that would be kind of strange. I can let one run for a long, long time and check if you think that I should.
  - QuickPlay Pro 2 uses this counter, but QuickPlay Pro 1 just set all four bytes to `0x00`. It must be optional, but I assume this makes the packets more robust.
  - I assume that it just overflows back to 0 when it gets to `0xffffffff`.
- The first byte after the port number is definitely some sort of value (not just padding).
  - I saw it change from `0x00` when the buffer size was not 512. It seems to change based on the size of the buffer, so maybe it is some sort of checksum? If so, I suppose it would never be used if we always send 512 bytes per buffer.
- I'm not sure what the next two bytes were, but they were always `0x0000`.
- The buffer size parameter is big-endian.
  - QuickPlay Pro 1 allowed me to set the buffer size sent, but QuickPlay Pro 2 always sends all 512 bytes.
  - In the settings panel of QuickPlay Pro 1, it says that a buffer can be a minimum of 24 bytes, but it seems that OLA always sends all 512 anyway.
- The next two bytes were always `0x0000`. I'm not sure if this is really a DMX START code or not, but it is definitely always `0x0000` from what I observed.

**I can confirm this works on a sPDS-60ca!**

DaAwesomeP committed:
fix kinet port out flag padding and bufer size, add packet number, add host to big endian conversions

DaAwesomeP commented:
OK, I have tested that this works with a PDM-202 as well! It seems that the packet counter is global across all devices in QuickPlay Pro 2, but I am not sure that it matters much.

DaAwesomeP committed:
kinet portout packet counter across all devices

peternewman commented on plugins/kinet/KiNetNode.h:
```diff
@@ -56,6 +56,7 @@ class KiNetNode {
 
  private:
     bool m_running;
+    static uint32_t m_packet_number;
```
We have a special type to handle this sort of thing:
http://docs.openlighting.org/ola/doc/latest/classola_1_1_sequence_number.html
See e.g.:
https://github.com/OpenLightingProject/ola/blob/c9add79a8d1c83ab209146006dae87906803b64f/libs/usb/JaRulePortHandleImpl.h#L112

DaAwesomeP replied:
Ah, ok I will switch it to that.

DaAwesomeP replied:
Fixed, see f8e2c10.

peternewman commented on plugins/kinet/KiNetNode.cpp:
```diff
@@ -177,9 +183,9 @@ void KiNetNode::SocketReady() {
   m_output_stream << KINET_MAGIC_NUMBER << KINET_VERSION_ONE;
-  m_output_stream << msg_type << sequence_number;
+  m_output_stream << msg_type << HostToBigEndian(m_packet_number);
+  m_packet_number++;
```
So just to check, this should also increment with a DMXOUT packet (for example)? But it's one global counter between all the different packet types, not one per packet type?

DaAwesomeP replied:
> But it's one global counter between all the different packet types, not one per packet type?

Yes. The counter appears to increment for each packet sent by QuickPlay Pro 2 no matter the packet type or device. I just did some more testing, and it seems that this sequence number is also used by end devices when responding to a packet. It seems that the end device responds with the same sequence number to match a response to a request.

>So just to check, this should also increment with a DMXOUT packet (for example)? 

Yes, I believe so. I don't know how to make QuickPlay Pro 2 produce a DMXOUT packet, so I cannot test that specifically. But every single packet put out by QuickPlay Pro 2 (not just PORTOUT) seems to have this counter in it.

peternewman replied:
> Yes. The counter appears to increment for each packet sent by QuickPlay Pro 2 no matter the packet type or device. I just did some more testing, and it seems that this sequence number is also used by end devices when responding to a packet.

Ah, so actually the sequence number will help us to make sense of the endian stuff, can you do a capture for when the sequence number rolls over from 0x00ff to 0x0100 and that should confirm the endianness of that counter too.

> It seems that the end device responds with the same sequence number to match a response to a request.

Ah, that was going to be my next question! So they reply with the same sequence number value.

> Yes, I believe so. I don't know how to make QuickPlay Pro 2 produce a DMXOUT packet, so I cannot test that specifically. But every single packet put out by QuickPlay Pro 2 (not just PORTOUT) seems to have this counter in it.

Ah, I thought I found a way to do that, but maybe I'm getting mixed up. The other packets confirm it though!

DaAwesomeP replied:
> Ah, so actually the sequence number will help us to make sense of the endian stuff, can you do a capture for when the sequence number rolls over from 0x00ff to 0x0100 and that should confirm the endianness of that counter too.

I compared side by side what my commits spit out and what QuickPlay Pro 2 spits out, and they currently match (I tested rolling the first 3 bytes). So that should tell you what it should be.

The order I am seeing in packet capture is:
`0x0801` `0x01000000` `0xffffffff`
`0x0801` `0x02000000` `0xffffffff`
...
`0x0801` `0xff000000` `0xffffffff`
`0x0801` `0x00010000` `0xffffffff`
`0x0801` `0x01010000` `0xffffffff`

peternewman commented on include/ola/network/NetworkUtils.h:
```diff
@@ -121,6 +121,36 @@ int16_t HostToLittleEndian(int16_t value);
  */
 int32_t HostToLittleEndian(int32_t value);
 
+/**
+ * @brief 8-bit unsigned host to big endian conversion.
+ */
+inline uint8_t HostToBigEndian(uint8_t value) { return value; }
+
+/**
+ * @brief 16-bit unsigned host to big endian conversion.
+ */
+uint16_t HostToBigEndian(uint16_t value);
+
+/**
+ * @brief 32-bit unsigned host to big endian conversion.
+ */
+uint32_t HostToBigEndian(uint32_t value);
+
+/**
+ * @brief 8-bit signed host to big endian conversion.
+ */
+inline int8_t HostToBigEndian(int8_t value) { return value; }
+
+/**
+ * @brief 16-bit signed host to big endian conversion.
+ */
+int16_t HostToBigEndian(int16_t value);
+
+/**
+ * @brief 32-bit signed host to big endian conversion.
+ */
+int32_t HostToBigEndian(int32_t value);
+
```
So we've got HostToNetwork functions which do this already.

DaAwesomeP replied:
Got it, will switch.

DaAwesomeP replied:
Fixed, see f8e2c10.

peternewman commented on plugins/kinet/KiNetNode.cpp:
```diff
@@ -125,8 +131,8 @@ bool KiNetNode::SendDMX(const IPV4Address &target_ip, const DmxBuffer &buffer) {
 bool KiNetNode::SendPortOut(const IPV4Address &target_ip,
                             const uint8_t port,
                             const DmxBuffer &buffer) {
-  static const uint8_t flags = 0;
-  static const uint8_t padding = 0;
+  static const uint8_t flags1 = 0; // Definitely flags of some sort, seems to be impacted by buffer size
+  static const uint16_t flags2 = 0; // Possibly always 0
```
So we've got some values for these:
https://github.com/OpenLightingProject/ola/blob/6cc16fe68d42a7edd4e273b89cf5cbbcbb530b7d/plugins/kinet/kinet.cpp#L154-L160

I've seen the 16 bit one with QPP2 at least, I'm not sure I triggered the sync one when I tried.

DaAwesomeP replied:
The power supplies that I have will not take in 16-bit data I don't think, so I have no way to test this. Possibly I can add a virtual power supply that supports 16-bit to QuickPlay Pro 2 to get it to spit it out.

DaAwesomeP replied:
OK, I can't get QuickPlay Pro 2 to do anything other than `0x000000` for all three flag bytes. It seems that is safe for most circumstances considering it was the same for my PDM-202 as well.

peternewman replied:
Yeah makes sense, we've got the comments in the other file should someone ever need it and have the ability to test it.

Maybe I saw it with my other manufacturer's device then, rather than QPP2.

peternewman commented on plugins/kinet/KiNetNode.cpp:
```diff
@@ -125,8 +131,8 @@ bool KiNetNode::SendDMX(const IPV4Address &target_ip, const DmxBuffer &buffer) {
                             const DmxBuffer &buffer) {
-  static const uint8_t flags = 0;
-  static const uint8_t padding = 0;
+  static const uint8_t flags1 = 0; // Definitely flags of some sort, seems to be impacted by buffer size
```
Could you be more specific how the buffer size impacts it?

DaAwesomeP replied:
I'm not particularly sure, but I will make a key-value table so we can figure it out.

DaAwesomeP replied:
See below, I'm no longer sure what this is, but it seems to be safe to leave at `0x00`.

peternewman commented on plugins/kinet/KiNetNode.cpp:
```diff
@@ -136,9 +142,9 @@ bool KiNetNode::SendPortOut(const IPV4Address &target_ip,
-                  << static_cast<uint16_t>(buffer.Size());
+  m_output_stream << universe << port
+                  << flags1 << flags2
+                  << HostToBigEndian(static_cast<uint16_t>(buffer.Size())); // Buffer must be at least 24 bytes
```
You can send a smaller packet with OLA like so:
`./examples/ola_streaming_client -u 2 --dmx 1,2,3,4`

The website only sends a full frame of 512 slots each time.

DaAwesomeP replied:
OK, I will add some code to pad with zeros if the buffer is smaller.

This comes from the QuickPlay Pro 1 options panel:
![image](https://user-images.githubusercontent.com/4472083/190932351-48cd47e8-1f8d-4b82-b0c3-3ceac8d4288b.png)

DaAwesomeP replied:
OK, I am now actually enforcing the minimum size. See 68d6dfe.

peternewman replied:
> This comes from the QuickPlay Pro 1 options panel: ![image](https://user-images.githubusercontent.com/4472083/190932351-48cd47e8-1f8d-4b82-b0c3-3ceac8d4288b.png)

So can you set the range in QPP1 to fewer than 24 channels? If so does it pad in QPP1, or in the PSU?

DaAwesomeP replied:
> So can you set the range in QPP1 to fewer than 24 channels? If so does it pad in QPP1, or in the PSU?

Yes. When QuickPlay Pro 1 is set to fewer than 24 channels, it always sends a buffer size of 24 and 24 data bytes. It fills in with zeros to hit the 24 data bytes. My latest commits do this as well.

This setting is also how I confirmed the endianness of the buffer size bytes. What my commits spit out matches what I saw in QuickPlay Pro 1 for the byte ordering.

peternewman reviewed (commented):
Thanks very much for this!

A few comments, some clarifying some bits, others pointing out some stuff we can drop/use our existing functionality.

peternewman commented:
> OK, after looking at the outputs of QuickPlay Pro 1 and QuickPlay Pro 2, I found some things:
> 
>     * I added the packet counter.

Thanks!

>       * It is a big-endian counter that increments for each packet sent. It is not a frame counter, as it increments for each port packet that is sent to a device.
>       * I don't know if it is a global counter across all devices for one KiNET source/generator or per-device, but I can test that next week. It probably doesn't matter since the devices wouldn't be able to tell anyway (unless it is used to check for missed packets and then it would need to be per-device).
>       * I saw this counter climb to at least 3 bytes, but it is probably 4 bytes since otherwise that would be kind of strange. I can let one run for a long, long time and check if you think that I should.

If it's per device that would be harder possibly, if it's not, then can't you add a few devices with multiple ports to generate even more packets per second?

>       * QuickPlay Pro 2 uses this counter, but QuickPlay Pro 1 just set all four bytes to `0x00`. It must be optional, but I assume this makes the packets more robust.

Well I assume the fixture can reject out of sequence packets.

>     * The first byte after the port number is definitely some sort of value (not just padding).

You've seen different values then? What sort of values? What did you do to cause them to happen?

>       * I saw it change from `0x00` when the buffer size was not 512. It seems to change based on the size of the buffer, so maybe it is some sort of checksum? If so, I suppose it would never be used if we always send 512 bytes per buffer.

What does it change to though, the slot count?

>     * The buffer size parameter is big-endian.

Our comment says it's little-endian:
https://github.com/OpenLightingProject/ola/blob/6cc16fe68d42a7edd4e273b89cf5cbbcbb530b7d/plugins/kinet/kinet.cpp#L174

Although it also says the whole protocol is LE:
https://github.com/OpenLightingProject/ola/blob/6cc16fe68d42a7edd4e273b89cf5cbbcbb530b7d/plugins/kinet/kinet.cpp#L44-L45

Which is sort of why I wanted to clarify some of the other stuff we can control like that (and possibly serial number if you they are numeric only).

I'm wondering if most of the protocol is BE, but that is LE for some reason, which would explain the issue, we're using a BigEndianOutputStream:
https://github.com/OpenLightingProject/ola/blob/6cc16fe68d42a7edd4e273b89cf5cbbcbb530b7d/plugins/kinet/KiNetNode.h#L61
http://docs.openlighting.org/ola/doc/latest/classola_1_1io_1_1_big_endian_output_stream.html

Which uses this:
http://docs.openlighting.org/ola/doc/latest/classola_1_1io_1_1_big_endian_output_stream_adaptor.html

Which ultimately writes via our HostToNetwork wrapper:
https://github.com/OpenLightingProject/ola/blob/c9add79a8d1c83ab209146006dae87906803b64f/include/ola/io/BigEndianStream.h#L154-L157

So if it's really BE, it should be just working. If it's actually LE, then your addition will convert it to BE, which our BigEndianOutputStream will then essentially convert to LE on the wire.

>     * I'm not sure what the next two bytes were, but they were always `0x0000`.

Isn't that the potential start code ones?

>       * QuickPlay Pro 1 allowed me to set the buffer size sent, but QuickPlay Pro 2 always sends all 512 bytes.

Ah perfect, could you try those three numbers I mentioned before for different buffer sizes then (which was what I was actually trying to work out)?

>       * In the settings panel of QuickPlay Pro 1, it says that a buffer can be a minimum of 24 bytes, but it seems that OLA always sends all 512 anyway.

As mentioned in the code review, OLA can send smaller data packets. 

>     * The next two bytes were always `0x0000`. I'm not sure if this is really a DMX START code or not, but it is definitely always `0x0000` from what I observed.

Our notes suggest it's different on chromASIC stuff (presumably actually Chromasic):
https://github.com/OpenLightingProject/ola/blob/6cc16fe68d42a7edd4e273b89cf5cbbcbb530b7d/plugins/kinet/kinet.cpp#L175

> **I can confirm this works on a sPDS-60ca!**

:partying_face:

peternewman commented:
It would also be great to improve/extend my tests for these new discoveries. I think I've just pushed one with the correct buffer size data now, but I guess if we test a second it should cover the counter stuff too...

DaAwesomeP commented:
> If it's per device that would be harder possibly, if it's not, then can't you add a few devices with multiple ports to generate even more packets per second?

See the second comment and commit; I confirmed that it is global for all packets produced from QuickPlay Pro 2. Yes, I can generate many packets at once and  the counter is incremented in the order in which they are sent.

> You've seen different values then? What sort of values? What did you do to cause them to happen?

I will try different buffer size values and make a table. **UPDATE:** I have no clue what it is doing, it is not just related to buffer size.

> What does it change to though, the slot count?

No, the slot count needs two bytes so it must be some kind of checksum I think.

> So if it's really BE, it should be just working. If it's actually LE, then your addition will convert it to BE, which our BigEndianOutputStream will then essentially convert to LE on the wire.

I'm definitely a bit confused by all of this...you're saying that there is code that I didn't see that is actually converting the big-endian back to little-endian? I could definitely be confusing which endian is which, but I definitely had to swap it as you see here.

> Isn't that the potential start code ones?

No, everything was shifted incorrectly. It goes:
1. Universe (`uint32_t`)
2. Port (`uint8_t`)
3. Unknown flags ~~related to buffer size~~ that I have no clue what it is doing (`uint8_t`)
4. Unknown flags that I only saw as `0x0000` (`uint16_t`)
5. Buffer size (`uint16_t`)
6. Possibly start code (`uint16_t`)
7. Data

> Ah perfect, could you try those three numbers I mentioned before for different buffer sizes then (which was what I was actually trying to work out)?

Will do.

> Our notes suggest it's different on chromASIC stuff (presumably actually Chromasic)

I am using Chromasic fixtures. Possibly this is related to when the power supply is receiving 16-bit data? Or maybe only certain power supplies use this field? I only saw `0x0000`.

DaAwesomeP commented:
Hmmm, OK I have no clue what the single byte `flag1` is doing.
- I am no longer so sure that it is related to buffer size. I saw it appear with the full 512 buffer.
- Only QuickPlay Pro 1 changes this field.
- QuickPlay Pro 2 never touches this field, so it seems to be safe to leave at `0x00`.

Here  is what I observed:

| buffer size | `flag1` |
|-|-|
|`0x18`|switches (seemingly randomly) between `0x48`, `0x16`, `0x3c`, and `0xca` even while the buffer is constant|
|`0x19`|switches between `0xc5`, `0x06`, `0x00`, `0x3c`, and `0x48`|
| |I tested some more combinations, but I did not write them down because it seems impossible to figure out|

This flag seemed to stay fixed at its current value unless I dragged around the color picker. Even if the DMX values stayed constant (i.e. moving from red to white for only one single red LED channel), this value would change in an unpredictable way when dragging the color picker. It is probably some sort of bit-flag and not a whole-byte number.

I thought maybe it would change in a certain order (like a sequence counter of some sort), but it does not.

We should just leave this byte at `0x00`.

DaAwesomeP committed:
kinet portout switch to ola existing big endian conversion and sequence numbering methods

DaAwesomeP committed:
kinet portout send minimum buffer size

DaAwesomeP commented:
OK, I fixed the few things for which I was not using OLA existing methods as well as added enforcement for the minimum packet size.

I gave up on figuring out the first flags byte after the port number, as it definitely appears to be optional and doesn't seem to follow any obvious pattern, sequence, or correlation to the buffer size or data sent.

For the next two bytes of flags, I only can get QuickPlay Pro 2 to output `0x0000`. I think it's probably safe to leave this for now. Same thing with the two bytes after the buffer size.

Let me know if I should change how I am handling the endianness. Though for a larger change it may be easier/faster for you to merge this pull, fix the underlying processing, and then squash it all in the other original pull request. What I can say for sure is that what I have committed here outputs working packets. These packets match what I have seen regardless of how many times OLA is actually flipping the endianness to land there.

DaAwesomeP commented:
> It would also be great to improve/extend my tests for these new discoveries. I think I've just pushed one with the correct buffer size data now, but I guess if we test a second it should cover the counter stuff too...

We can also now test that the buffer is always 24 bytes and that it pads the end with zeros if needed to get there.

peternewman commented:
> See the second comment and commit; I confirmed that it is global for all packets produced from QuickPlay Pro 2. Yes, I can generate many packets at once and the counter is incremented in the order in which they are sent.

Cool, I was more thinking you might get to the fourth byte rollover a bit quicker! Although at 40 FPS with only one port it should "only" take 27 minutes by my maths, which isn't too horrendous, and would be a reported issue if it went wrong at that point, although I appreciate the three years for the full rollover is unreasonable to test.

> No, the slot count needs two bytes so it must be some kind of checksum I think.

👍 

> I'm definitely a bit confused by all of this...you're saying that there is code that I didn't see that is actually converting the big-endian back to little-endian? I could definitely be confusing which endian is which, but I definitely had to swap it as you see here.

Yeah, as mentioned here we're doing a HostToNetwork conversion already:
https://github.com/OpenLightingProject/ola/blob/c9add79a8d1c83ab209146006dae87906803b64f/include/ola/io/BigEndianStream.h#L154-L157

The general issue is apart from the counter, and the buffer size, i.e. the stuff we control, it's impossible to tell the endianness of the protocol because the packet type constants for example, we don't know which way round they actually should be!

> > Our notes suggest it's different on chromASIC stuff (presumably actually Chromasic)
> 
> I am using Chromasic fixtures. Possibly this is related to when the power supply is receiving 16-bit data? Or maybe only certain power supplies use this field? I only saw `0x0000`.

Or a different type of Chromasic stuff?

DaAwesomeP commented:
> Cool, I was more thinking you might get to the fourth byte rollover a bit quicker! Although at 40 FPS with only one port it should "only" take 27 minutes by my maths, which isn't too horrendous, and would be a reported issue if it went wrong at that point, although I appreciate the three years for the full rollover is unreasonable to test.

OK, I'll leave it running all day today and report back!

> Yeah, as mentioned here we're doing a HostToNetwork conversion already:
https://github.com/OpenLightingProject/ola/blob/c9add79a8d1c83ab209146006dae87906803b64f/include/ola/io/BigEndianStream.h#L154-L157
>
> The general issue is apart from the counter, and the buffer size, i.e. the stuff we control, it's impossible to tell the endianness of the protocol because the packet type constants for example, we don't know which way round they actually should be!

I see. Well, what I can say is that the packets I am producing here seem to be correct and match the QuickPlay Pro ones. Whether or not we convert these parts and then flip the whole thing or just flip some other parts is an implementation detail that I'll leave to you.

> Or a different type of Chromasic stuff?

I think all of the Chromasic stuff is the same, it's just the power supplies that drive them that differ. Regardless, I do not have a power supply that supports the 16-bit mode that might be referring to.

peternewman commented:
> OK, I'll leave it running all day today and report back!

Amazing thanks!

> I see. Well, what I can say is that the packets I am producing here seem to be correct and match the QuickPlay Pro ones. Whether or not we convert these parts and then flip the whole thing or just flip some other parts is an implementation detail that I'll leave to you.

Yeah fair point, as you said before, getting the tests in and working is probably most important, the rest is then decoration on top! On which note would you mind fixing the tests now? Initially just by adding the new padding?
https://app.travis-ci.com/github/peternewman/ola/jobs/583157759#L5044-L5058

There's also some minor lint issues:
https://app.travis-ci.com/github/peternewman/ola/jobs/583157751#L1588-L1595

> > Or a different type of Chromasic stuff?
> 
> I think all of the Chromasic stuff is the same, it's just the power supplies that drive them that differ. Regardless, I do not have a power supply that supports the 16-bit mode that might be referring to.

Fair enough, although I think that was primarily supposedly a different start code...

DaAwesomeP committed:
Merge remote-tracking branch 'peternewman/kinet-port-out' into kinet-port-out

DaAwesomeP committed:
kinet port out fix lint

DaAwesomeP committed:
fix and comment kinet port out test, comment dmx out test

DaAwesomeP commented:
> > OK, I'll leave it running all day today and report back!
> 
> Amazing thanks!

Well, I think it will need another half-day (I think 2.5 days total?). The frame rate of QuickPlay Pro 2 is not very fast it seems, although I know these devices can work up to at least 60 fps. The third byte is up to `0xaf`.

> Yeah fair point, as you said before, getting the tests in and working is probably most important, the rest is then decoration on top! On which note would you mind fixing the tests now? Initially just by adding the new padding?
> https://app.travis-ci.com/github/peternewman/ola/jobs/583157759#L5044-L5058

Fixed in d35d4cd.

> There's also some minor lint issues:
> https://app.travis-ci.com/github/peternewman/ola/jobs/583157751#L1588-L1595

Fixed in c54fa8d.

peternewman commented on plugins/kinet/KiNetNode.cpp:
```diff
@@ -18,23 +18,32 @@
 using ola::network::UDPSocket;
 using std::auto_ptr;
 
+/*
+ * Initialize packet counter.
+ */
+ola::SequenceNumber<uint32_t> KiNetNode::m_transaction_number = (
+  ola::SequenceNumber<uint32_t>());
+
```
You can drop this and just do it in the header:
```suggestion
```

DaAwesomeP replied:
It's a static variable, so I was getting errors if I did not define it outside the header.

DaAwesomeP replied:
Fixed!

peternewman commented on plugins/kinet/KiNetNode.h:
```diff
@@ -56,6 +57,7 @@ class KiNetNode {
 
  private:
     bool m_running;
+    static ola::SequenceNumber<uint32_t> m_transaction_number;
```
Unless your static is upsetting that? Although arguably two nodes would be the same as two copies of Quick Play Pro, which would presumably produce two sets of transaction numbers?

DaAwesomeP replied:
Yes, the static was why it needed to be defined separately.

No, when I had two power supplies configured at the same time in one QuickPlay Pro 2 instance it incremented the packet counter globally across all of them. So similarly every packet sent by OLA should be incremented.

peternewman replied:
> No, when I had two power supplies configured at the same time in one QuickPlay Pro 2 instance it incremented the packet counter globally across all of them. So similarly every packet sent by OLA should be incremented.

Yeah but I was asking about two copies of QPP2. The KiNetNode is defined in the plugin so should only exist once (although admittedly I haven't tested that yet...).

DaAwesomeP replied:
> Yeah but I was asking about two copies of QPP2. The KiNetNode is defined in the plugin so should only exist once (although admittedly I haven't tested that yet...).

Ah, so there is only one instance of the `KiNetNode` class? In that case this indeed does not need to be static.

DaAwesomeP replied:
Well, at least in the tests it is spawned twice, but that is not a big deal. In the plugin it does appear to only be spawned once. It looks like if `StartHook()` is called more than once then it might spawn a second, but I think that's a good opportunity for the counter to reset anyway. I'll fix it.

DaAwesomeP replied:
Fixed!

peternewman commented on plugins/kinet/KiNetNode.cpp:
```diff
@@ -125,23 +134,33 @@ bool KiNetNode::SendDMX(const IPV4Address &target_ip, const DmxBuffer &buffer) {
     return true;
   }
 
+  uint16_t buffSize = static_cast<uint16_t>(buffer.Size());
+  uint16_t buffSizeRegulated = std::max(
+    buffSize,
+    KINET_PORTOUT_MIN_BUFFER_SIZE);
```
Minor style tweaks:
```suggestion
  uint16_t buffer_size = static_cast<uint16_t>(buffer.Size());
  uint16_t buffer_size_regulated = std::max(
    buffer_size,
    KINET_PORTOUT_MIN_BUFFER_SIZE);
```

DaAwesomeP replied:
Fixed.

peternewman commented on plugins/kinet/KiNetNode.cpp:
```diff
@@ -125,23 +134,33 @@ bool KiNetNode::SendDMX(const IPV4Address &target_ip, const DmxBuffer &buffer) {
-                  << static_cast<uint16_t>(buffer.Size());
+  m_output_stream << universe << port
+                  << flags1 << flags2
+                  << HostToNetwork(buffSizeRegulated);
```
Matching code style:
```suggestion
                  << HostToNetwork(buffer_size_regulated);
```

DaAwesomeP replied:
Fixed.

peternewman commented on plugins/kinet/KiNetNode.cpp:
```diff
@@ -125,23 +134,33 @@ bool KiNetNode::SendDMX(const IPV4Address &target_ip, const DmxBuffer &buffer) {
   m_output_stream.Write(buffer.GetRaw(), buffer.Size());
 
+  // Buffer must be at least 24 bytes, pad with zeros if needed
+  for (uint16_t i = 0; i < (KINET_PORTOUT_MIN_BUFFER_SIZE - buffSize); i++) {
```
Do you think this is more readable (assuming it's still actually accurate), just a genuine question?
```suggestion
  for (uint16_t i = buffer_size; i <= KINET_PORTOUT_MIN_BUFFER_SIZE; i++) {
```

DaAwesomeP replied:
I think it should still be `<`, not `<=`, but sure.

DaAwesomeP replied:
Fixed.

peternewman replied:
Yeah I wasn't certain, but than buffer_size is 1 based essentially. Actually you are probably right, otherwise a buffer of 24 would still append a zero.

```suggestion
  for (uint16_t i = buffer_size; i < KINET_PORTOUT_MIN_BUFFER_SIZE; i++) {
```

peternewman commented on plugins/kinet/KiNetNode.cpp:
```diff
@@ -125,23 +134,33 @@ bool KiNetNode::SendDMX(const IPV4Address &target_ip, const DmxBuffer &buffer) {
     return true;
   }
 
+  uint16_t buffSize = static_cast<uint16_t>(buffer.Size());
+  uint16_t buffSizeRegulated = std::max(
+    buffSize,
+    KINET_PORTOUT_MIN_BUFFER_SIZE);
+
   m_output_queue.Clear();
   PopulatePacketHeader(KINET_PORTOUT_MSG);
-  m_output_stream << universe << port << padding
-                  << flags << padding  // Are the flags actually 16 bit?
-                  << static_cast<uint16_t>(buffer.Size());
+  m_output_stream << universe << port
+                  << flags1 << flags2
+                  << HostToNetwork(buffSizeRegulated);
   m_output_stream << static_cast<uint16_t>(DMX512_START_CODE);
   m_output_stream.Write(buffer.GetRaw(), buffer.Size());
 
+  // Buffer must be at least 24 bytes, pad with zeros if needed
```
```suggestion
  m_output_stream.Write(buffer.GetRaw(), buffer.Size());

  // TODO(Peter): Update to our new DmxBuffer padding options when we add and write them.
  // Buffer must be at least 24 bytes, pad with zeros if needed
```

DaAwesomeP replied:
Fixed.

peternewman reviewed (changes requested):
LGTM thanks, just some minor stylistic tweaks...

peternewman commented:
> > > OK, I'll leave it running all day today and report back!
> > 
> > 
> > Amazing thanks!
> 
> Well, I think it will need another half-day (I think 2.5 days total?). The frame rate of QuickPlay Pro 2 is not very fast it seems, although I know these devices can work up to at least 60 fps. The third byte is up to `0xaf`.

:open_mouth: I didn't realise it would be that slow, that's very dedicated!

Can you change the universe number in either version of QPP? I figure that might provide the conclusive answer regarding protocol endianness...?

DaAwesomeP commented:
> open_mouth I didn't realise it would be that slow, that's very dedicated!

It's still going! I'll let you know tomorrow.

> Can you change the universe number in either version of QPP? I figure that might provide the conclusive answer regarding protocol endianness...?

No, it always outputs `0xffffffff` for PORTOUT packets and I cannot get it to output DMXOUT packets. I can send a packet that _sets_ the universe (but doesn't send universe data), but I don't think that is helpful.

DaAwesomeP commented:
> > Can you change the universe number in either version of QPP? I figure that might provide the conclusive answer regarding protocol endianness...?
> 
> No, it always outputs `0xffffffff` for PORTOUT packets and I cannot get it to output DMXOUT packets. I can send a packet that _sets_ the universe (but doesn't send universe data), but I don't think that is helpful.

Oh, actually I see what you mean now. I suppose they would not flip the endianness between packets. I'll try the set universe packet tomorrow.

DaAwesomeP committed:
kinet port out style fixes

DaAwesomeP committed:
kinet port out make counter not static

DaAwesomeP committed:
kinet fix universe endianness

DaAwesomeP committed:
Merge branch 'kinet-port-out' of github.com:DaAwesomeP/ola into kinet-port-out

DaAwesomeP commented:
> > open_mouth I didn't realise it would be that slow, that's very dedicated!
> 
> It's still going! I'll let you know tomorrow.

Alright! I have confirmed it is indeed a 4-byte counter! It finally made it to `0x01` for the fourth byte.

> Oh, actually I see what you mean now. I suppose they would not flip the endianness between packets. I'll try the set universe packet tomorrow.

The packet used to set the universe option of the device looks like this from QuickPlay Pro 2:
|magic number|packet type|counter|unknown flags|universe|universe in UI|
|-|-|-|-|-|-|
|`0x04 0x01 0xdc 0x4a 0x01 0x00`|`0x05 0x00`|`0xe4 0xdb 0x32 0x01`|`0xef 0xbe 0xad 0xde`|`0x00 0x00 0x00 0x00`|0|
|`0x04 0x01 0xdc 0x4a 0x01 0x00`|`0x05 0x00`|`0xe5 0xdb 0x32 0x01`|`0xef 0xbe 0xad 0xde`|`0x01 0x00 0x00 0x00`|1|
|`0x04 0x01 0xdc 0x4a 0x01 0x00`|`0x05 0x00`|`0xe6 0xdb 0x32 0x01`|`0xef 0xbe 0xad 0xde`|`0xff 0x00 0x00 0x00`|255|
|`0x04 0x01 0xdc 0x4a 0x01 0x00`|`0x05 0x00`|`0xe7 0xdb 0x32 0x01`|`0xef 0xbe 0xad 0xde`|`0x00 0x01 0x00 0x00`|256|
|`0x04 0x01 0xdc 0x4a 0x01 0x00`|`0x05 0x00`|`0xe8 0xdb 0x32 0x01`|`0xef 0xbe 0xad 0xde`|`0x01 0x04 0x00 0x00`|1025|
|`0x04 0x01 0xdc 0x4a 0x01 0x00`|`0x05 0x00`|`0xe9 0xdb 0x32 0x01`|`0xef 0xbe 0xad 0xde`|`0xfe 0xff 0x00 0x00`|65534|
|`0x04 0x01 0xdc 0x4a 0x01 0x00`|`0x05 0x00`|`0xea 0xdb 0x32 0x01`|`0xef 0xbe 0xad 0xde`|`0xff 0xff 0x00 0x00`|65535|

QuickPlay Pro 2 did not allow me to set a universe higher than 65535, and QuickPlay Pro 1 only allows for up to 255.

I also confirmed again that the counter increments across all packet types, as these is the same counter values I was just seeing in my extended run.

Considering that the command to set the universe is specifically used in combination with the DMXOUT packets, I think this also more solidly confirms that DMXOUT must employ the packet counter as well.

I pushed another commit fixing the universe endianness, though we always send `0xffffffff` for now.

DaAwesomeP commented:
Build is passing! Since we confirmed the number of bytes of the counter and the endianness of the universe, is that everything?

DaAwesomeP commented:
Force-pushed to fix a GPG signing issue on the merge.

peternewman reviewed (approved):
Thanks for all your work on this @DaAwesomeP ! I certainly couldn't have done some of this without a test fixture.

For future info, you can merge pull request code suggestions directly like so:
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

Which is probably easier to track and also credits the author of them too (not that I'm too worried about that).

Now we can get onto the fun bit of how we actually present this in OLA back in the original PR!

DaAwesomeP commented:
Thank you! I'm really excited for this.

If you squash the commits in the other pull through the GitHub interface, then it will attribute the squashed commit to multiple people if there are commits from multiple people. The suggest changes feature works similarly.

https://github.blog/changelog/2019-12-19-improved-attribution-when-squashing-commits/

I'm not sure what the default behavior of the Git command line is (I have not tried squashing commits from multiple people from there).
